controller/src/main/java/org/jboss/as/controller/CapabilityReferenceRecorder.java

```diff
@@ -186,34 +232,42 @@ public final void addCapabilityRequirements(OperationContext context, String att
          * @throws AssertionError if the requirements discussed in the class javadoc are not fulfilled
          */
         @Override
-        public final void removeCapabilityRequirements(OperationContext context, String attributeName, String... attributeValues) {
-            processCapabilityRequirement(context, attributeName, true, attributeValues);
+        public final void removeCapabilityRequirements(OperationContext context, Resource resource, String attributeName, String... attributeValues) {
+            processCapabilityRequirement(context, resource, attributeName, true, attributeValues);
         }
 
-        private void processCapabilityRequirement(OperationContext context, String attributeName, boolean remove, String... attributeValues) {
-            String dependentName = getDependentName(context);
-
+        protected void processCapabilityRequirement(OperationContext context, Resource resource, String attributeName, boolean remove, String... attributeValues) {
+            RuntimeCapability cap = getDependentCapability(context);
+            String dependentName = getDependentName(cap, context);
             for (String attributeValue : attributeValues) {
-                String requirementName = RuntimeCapability.buildDynamicCapabilityName(baseRequirementName, attributeValue);
-                if (remove) {
-                    context.deregisterCapabilityRequirement(requirementName, dependentName);
-                } else {
-                    context.registerAdditionalCapabilityRequirement(requirementName, dependentName, attributeName);
+                if (attributeValue != null || !cap.getDynamicOptionalRequirements().contains(baseRequirementName)) {
```

Comment:
How can null ever be a useful value here?

If the AD allows undefined, since no one ever populates RuntimeCap's dynamic optional requirements, this will just result in registering a strange requirement.